File: backend/aegis/models/thesis.py
"""ThesisCard — Episodic Memory: complete thesis lifecycle record."""

from datetime import datetime
from typing import Any

from sqlalchemy import JSON, Boolean, DateTime, Float, Integer, String, func
from sqlalchemy.orm import Mapped, mapped_column

from aegis.models.base import Base


class ThesisCard(Base):
    __tablename__ = "thesis_cards"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    ticker: Mapped[str] = mapped_column(String(20), index=True, nullable=False)
    direction: Mapped[str] = mapped_column(String(20), nullable=False)
    entry_mode: Mapped[str] = mapped_column(String(20), nullable=False)
    entry_date: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    entry_price: Mapped[float] = mapped_column(Float, nullable=False)
    target_price: Mapped[float | None] = mapped_column(Float, nullable=True)
    stop_price: Mapped[float | None] = mapped_column(Float, nullable=True)
    key_assumptions: Mapped[dict[str, Any]] = mapped_column(JSON, default=list)
    thesis_valid_status: Mapped[str] = mapped_column(String(20), default="valid")
    re_entry_flagged: Mapped[bool] = mapped_column(Boolean, default=False)
    factor_snapshot: Mapped[dict[str, Any]] = mapped_column(JSON, default=dict)
    close_date: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    close_price: Mapped[float | None] = mapped_column(Float, nullable=True)
    actual_pnl_pct: Mapped[float | None] = mapped_column(Float, nullable=True)
    judgment_score: Mapped[int | None] = mapped_column(Integer, nullable=True)
    execution_score: Mapped[int | None] = mapped_column(Integer, nullable=True)
    close_reason: Mapped[str | None] = mapped_column(String(100), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, server_default=func.now(), onupdate=func.now()
    )

File: frontend/lib/api/thesis.ts
const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000/api/v1";

export interface ThesisCard {
  id: number;
  ticker: string;
  direction: string;
  entry_mode: string;
  thesis_valid_status: string;
  entry_date: string | null;
  entry_price: number;
  target_price: number | null;
  stop_price: number | null;
  actual_pnl_pct: number | null;
  judgment_score: number | null;
  execution_score: number | null;
}

export interface ThesisDetail extends ThesisCard {
  key_assumptions: string[];
  factor_snapshot: Record<string, number>;
  close_date: string | null;
  close_price: number | null;
  close_reason: string | null;
  re_entry_flagged: boolean;
}

export async function fetchTheses(
  filters: Record<string, string>
): Promise<ThesisCard[]> {
  const params = new URLSearchParams();
  if (filters.status && filters.status !== "all") params.set("status", filters.status);
  if (filters.direction && filters.direction !== "all") params.set("direction", filters.direction);
  if (filters.ticker) params.set("ticker", filters.ticker);

  const res = await fetch(`${API_BASE}/thesis?${params}`);
  if (!res.ok) throw new Error(`API ${res.status}`);
  const data = await res.json();
  return data.items;
}

export async function fetchThesisDetail(id: number): Promise<ThesisDetail> {
  const res = await fetch(`${API_BASE}/thesis/${id}`);
  if (!res.ok) throw new Error(`API ${res.status}`);
  return res.json();
}

export async function closeThesis(
  id: number,
  data: {
    close_price: number;
    judgment_score: number;
    execution_score: number;
    close_reason: string;
  }
): Promise<Record<string, unknown>> {
  const res = await fetch(`${API_BASE}/thesis/${id}/close`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail || `API ${res.status}`);
  }
  return res.json();
}